src/Confluent.Kafka.Avro/AvroProducer.cs

```diff
@@ -0,0 +1,349 @@
+        ///     The type of serializer to use to serialize keys.
+        /// </param>
+        /// <param name="valueSerdeType">
+        ///     The type of serializer to use to serialize values.
+        /// </param>
+        /// <param name="cancellationToken">
+        ///     A cancellation token that can be used to abort this request.
+        /// </param>
+        /// <returns>
+        ///     A <see cref="System.Threading.Tasks.Task" /> which will complete with 
+        ///     a delivery report corresponding to the produce request, or an exception
+        ///     if an error occured.
+        /// </returns>
+        public async Task<DeliveryReport<TKey, TValue>> ProduceAsync<TKey, TValue>(
+            TopicPartition topicPartition,
+            Message<TKey, TValue> message,
+            SerdeType keySerdeType,
+            SerdeType valueSerdeType,
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            try
+            {
+                var dr = await ProduceAsync(
+                    topicPartition,
+                    await CreateMessage(message, topicPartition.Topic, keySerdeType, valueSerdeType),
+                    cancellationToken
+                );
+
+                return new DeliveryReport<TKey, TValue>
+                {
+                    TopicPartitionOffset = dr.TopicPartitionOffset,
+                    Message = new Message<TKey, TValue>
+                    {
+                        Key = message.Key,
+                        Value = message.Value,
+                        Timestamp = dr.Timestamp,
+                        Headers = dr.Headers
+                    }
+                };
+            }
+            catch (AggregateException e)
+            {
+                throw e.InnerException;
+            }
+        }
+
+
+        /// <summary>
+        ///     Asynchronously produce a single
+        ///     <see cref="Confluent.Kafka.Message{TKey, TValue}" /> to a Kafka topic.
+        /// </summary>
+        /// <param name="topic">
+        ///     The topic to produce the message to.
+        /// </param>
+        /// <param name="message">
+        ///     The message to produce.
+        /// </param>
+        /// <param name="keySerdeType">
+        ///     The type of serializer to use to serialize keys.
+        /// </param>
+        /// <param name="valueSerdeType">
+        ///     The type of serializer to use to serialize values.
+        /// </param>
+        /// <param name="deliveryHandler">
+        ///     A delegate that will be called with a delivery report corresponding
+        ///     to the produce request (if enabled).
+        /// </param>
+        public void BeginProduce<TKey, TValue>(
+            string topic,
+            Message<TKey, TValue> message,
+            SerdeType keySerdeType,
+            SerdeType valueSerdeType,
+            Action<DeliveryReportResult<TKey, TValue>> deliveryHandler = null)
```

Comment:
we need two different names since the type is different to `DeliveryReport` as returned by `ProduceAsync` (includes `Error`). Since the result of a `Task` is referred to as `Result`, appending this to the name made sense to me. But I agree, it's not ideal. Open to other suggestions.

Reply:
how about just `DeliveryResult`?

Reply:
Very very good. But `DeliveryReportResult` shouldn't be called this - `DeliveryReport` should, and `DeliveryReportResult` named `DeliveryReport`. This is nice - as an added bonus it brings consistency with `ConsumeResult` (which also doesn't include an error). The word report (consistent with librdkafka and other clients) implies inclusion of an error. Use of Result doesn't (error is via an exception).

Reply:
i'm also including the change suggested by @AndyPook in the commit for this to inherit non generic classes from `<byte[], byte[]>` variants.